delta-app/src/main/java/io/cdap/delta/store/DBReplicationStateStore.java

```diff
@@ -0,0 +1,140 @@
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A StructuredTable based database table to store generic replicator states.
+ */
+public class DBReplicationStateStore {
+
+  private static final StructuredTableId TABLE_ID = new StructuredTableId("delta_state_store");
+  private static final String NAMESPACE_COL = "namespace";
+  private static final String GENERATION_COL = "generation";
+  private static final String APPNAME_COL = "app_name";
+  private static final String INSTANCEID_COL = "instance_id";
+  private static final String STATE_COL = "state";
+  private static final String VALUE_COL = "value";
+  private static final String UPDATED_COL = "last_updated";
+
+  public static final StructuredTableSpecification TABLE_SPEC = new StructuredTableSpecification.Builder()
+    .withId(TABLE_ID)
+    .withFields(new FieldType(NAMESPACE_COL, FieldType.Type.STRING),
+                new FieldType(GENERATION_COL, FieldType.Type.LONG),
+                new FieldType(APPNAME_COL, FieldType.Type.STRING),
+                new FieldType(INSTANCEID_COL, FieldType.Type.INTEGER),
+                new FieldType(STATE_COL, FieldType.Type.STRING),
+                new FieldType(VALUE_COL, FieldType.Type.STRING),
+                new FieldType(UPDATED_COL, FieldType.Type.LONG))
+    .withPrimaryKeys(NAMESPACE_COL, APPNAME_COL, GENERATION_COL, INSTANCEID_COL, STATE_COL)
+    .build();
+
+  private final StructuredTable table;
+
+  private DBReplicationStateStore(StructuredTable table) {
+    this.table = table;
+  }
+
+  static DBReplicationStateStore get(StructuredTableContext context) {
+    try {
+      StructuredTable table = context.getTable(TABLE_ID);
+      return new DBReplicationStateStore(table);
+    } catch (TableNotFoundException e) {
+      throw new IllegalStateException(String.format(
+        "System table '%s' does not exist. Please check your system environment.", TABLE_ID.getName()), e);
+    }
+  }
+
+  public String getStateData(String namespace, String appName, Long generation, int instanceId, String state)
+    throws IOException {
+    List<Field<?>> keys = getKey(namespace, appName, generation, instanceId, state);
+    Optional<StructuredRow> row = table.read(keys);
+    if (!row.isPresent() || row.get().getString(VALUE_COL) == null) {
+      return StringUtils.EMPTY;
+    }
+    return row.get().getString(VALUE_COL);
+  }
+
+  public void writeStateData(String namespace, String appName, Long generation, int instanceId, String state,
+                             String data) throws IOException {
+    Collection<Field<?>> fields = getKey(namespace, appName, generation, instanceId, state);
+    fields.add(Fields.stringField(VALUE_COL, data));
+    fields.add(Fields.longField(UPDATED_COL, System.currentTimeMillis()));
+
+    table.upsert(fields);
+  }
+
+  public Long getLatestGeneration(String namespace, String appName) throws IOException {
```

Comment:
why is this needed?

Reply:
if its for unit test, annotate it with `@VisibleForTesting`. It should also usually be package private instead of public

Reply:
No this one is used by the `getState` API [ assessment handler ] to track the status of the whole pipeline. 

Sorry, I misread the filename earlier